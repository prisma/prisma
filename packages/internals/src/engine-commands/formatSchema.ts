import fs from 'fs'
import { match } from 'ts-pattern'

import { logger } from '..'
import { ErrorArea, RustPanic } from '../panic'
import { prismaFmt } from '../wasm'
import { getLintWarningsAsText, lintSchema } from './lintSchema'

type FormatSchemaParams = { schema: string; schemaPath?: never } | { schema?: never; schemaPath: string }

function isSchemaOnly(schemaParams: FormatSchemaParams): schemaParams is { schema: string; schemaPath?: never } {
  return Boolean(schemaParams.schema)
}

function isSchemaPathOnly(schemaParams: FormatSchemaParams): schemaParams is { schema?: never; schemaPath: string } {
  return Boolean(schemaParams.schemaPath)
}

/**
 * Can be used by passing either the `schema` as a string, or a path `schemaPath` to the schema file.
 * Currently, we only use `schemaPath` in the cli. Do we need to keep supporting `schema` as well?
 */
export async function formatSchema(
  { schemaPath, schema }: FormatSchemaParams,
  inputFormattingOptions?: Partial<DocumentFormattingParams['options']>,
): Promise<string> {
  if (!schema && !schemaPath) {
    throw new Error(`Parameter schema or schemaPath must be passed.`)
  }

  if (process.env.FORCE_PANIC_PRISMA_FMT) {
    handleFormatPanic(
      () => {
        prismaFmt.debug_panic()
      },
      { schemaPath, schema } as FormatSchemaParams,
    )
  }

  const schemaContent = match({ schema, schemaPath } as FormatSchemaParams)
    .when(isSchemaOnly, ({ schema: _schema }) => _schema)
    .when(isSchemaPathOnly, ({ schemaPath: _schemaPath }) => {
      if (!fs.existsSync(_schemaPath)) {
        throw new Error(`Schema at ${schemaPath} does not exist.`)
      }

      const _schema = fs.readFileSync(_schemaPath, { encoding: 'utf8' })
      return _schema
    })
    .exhaustive()

  const defaultFormattingOptions: DocumentFormattingParams['options'] = {
    tabSize: 2,
    insertSpaces: true,
  }

  const documentFormattingParams = {
    textDocument: { uri: 'file:/dev/null' },
    options: {
      ...defaultFormattingOptions,
      ...inputFormattingOptions,
    },
  } as DocumentFormattingParams

  /**
   * Note:
   * - Given an invalid schema, `formatWasm` returns a formatted schema regardless (when it doesn't panic).
   * - Given an invalid schema, `lintSchema` returns a list of warnings/errors regardless (when it doesn't panic).
   *   Warnings must be filtered out from the other diagnostics.
   * - Validation errors aren't checked/shown here.
   *   They appear when calling `getDmmf` on the formatted schema in Format.ts.
   *   If we called `getConfig` instead, we wouldn't have any validation check.
   */
  const { formattedSchema, lintDiagnostics } = handleFormatPanic(
    () => {
      // the only possible error here is a Rust panic
      const formattedSchema = formatWasm(schemaContent, documentFormattingParams)
      const lintDiagnostics = lintSchema({ schema: formattedSchema })
      return { formattedSchema, lintDiagnostics }
    },
    { schemaPath, schema } as FormatSchemaParams,
  )

  const lintWarnings = getLintWarningsAsText(lintDiagnostics)
  if (lintWarnings && logger.should.warn()) {
    // Output warnings to stderr
    console.warn(lintWarnings)
  }

  return Promise.resolve(formattedSchema)
}

function handleFormatPanic<T>(tryCb: () => T, { schemaPath, schema }: FormatSchemaParams) {
  try {
    return tryCb()
  } catch (e: unknown) {
    const wasmError = e as Error
    throw new RustPanic(
      /* message */ wasmError.message,
      /* rustStack */ wasmError.stack || 'NO_BACKTRACE',
      /* request */ '@prisma/prisma-fmt-wasm format',
      ErrorArea.FMT_CLI,
      schemaPath,
      schema,
    )
  }
}

type DocumentUri = string

type TextDocument = {
  /**
   * The associated URI for this document. Most documents have the __file__-scheme, indicating that they
   * represent files on disk. However, some documents may have other schemes indicating that they are not
   * available on disk.
   */
  readonly uri: DocumentUri
}

// Part of the LSP spec: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentFormattingParams
// These are only the parts we are interested in.
type DocumentFormattingParams = {
  textDocument: TextDocument
  options: {
    // this is the only property currently considered by Rust, the rest are ignored but are needed for successfully unmarshaling the `DocumentFormattingParams`
    // and be compatible with the LSP spec.
    // The Wasm formatter may fail silently on unmarshaling errors (a `warn!` macro is used in the Rust code, but that's not propagated to Wasm land).
    tabSize: number
    insertSpaces: boolean
  }
}

function formatWasm(schema: string, documentFormattingParams: DocumentFormattingParams): string {
  const formattedSchema = prismaFmt.format(schema, JSON.stringify(documentFormattingParams))
  return formattedSchema
}
