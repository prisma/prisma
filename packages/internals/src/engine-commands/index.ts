export { formatSchema } from './formatSchema'
export type { ConfigMetaFormat } from './getConfig'
export { getConfig, getEffectiveUrl } from './getConfig'
export type { GetDMMFOptions } from './getDmmf'
export { getDMMF } from './getDmmf'
export { getEnginesMetaInfo } from './getEnginesMetaInfo'
export { getEngineVersion } from './getEngineVersion'
export { getLintWarningsAsText, handleLintPanic, lintSchema } from './lintSchema'
