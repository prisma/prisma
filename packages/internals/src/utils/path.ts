import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
