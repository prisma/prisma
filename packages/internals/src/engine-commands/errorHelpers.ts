import { formatTable } from '../utils/formatTable'
import { version } from '../utils/getVersionFromPackageJson'

/**
 * Adds `Prisma CLI Version : x.x.x` at the bottom of the error output.
 */
export function addVersionDetailsToErrorMessage(message: string) {
  const rows = [['Prisma CLI Version', version]]
  return `${message}

${formatTable(rows)}`
}
