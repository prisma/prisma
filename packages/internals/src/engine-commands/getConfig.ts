import Debug from '@prisma/debug'
import { getCliQueryEngineBinaryType } from '@prisma/engines'
import { BinaryType } from '@prisma/fetch-engine'
import type { DataSource, EnvValue, GeneratorConfig } from '@prisma/generator-helper'
import chalk from 'chalk'
import execa from 'execa'
import * as E from 'fp-ts/Either'
import { identity, pipe } from 'fp-ts/lib/function'
import * as TE from 'fp-ts/TaskEither'
import { match, P } from 'ts-pattern'

import { ErrorArea, isExecaErrorCausedByRustPanic, isWasmPanic, RustPanic } from '../panic'
import { prismaFmt } from '../wasm'
import { addVersionDetailsToErrorMessage } from './errorHelpers'
import {
  createDebugErrorType,
  createSchemaValidationError,
  loadNodeAPILibrary,
  preliminaryBinaryPipeline,
  preliminaryNodeAPIPipeline,
  unlinkTempDatamodelPath,
} from './queryEngineCommons'

const debug = Debug('prisma:getConfig')

const MAX_BUFFER = 1_000_000_000

export interface ConfigMetaFormat {
  datasources: DataSource[] | []
  generators: GeneratorConfig[] | []
  warnings: string[] | []
}

export type GetConfigOptions = {
  datamodel: string
  cwd?: string
  prismaPath?: string
  datamodelPath?: string
  retry?: number
  ignoreEnvVarErrors?: boolean
}

type GetConfigErrorInit = {
  // e.g., `Schema parsing - Error while interacting with query-engine-node-api library`
  reason: string

  // e.g., Error validating model "public": The model name \`public\` is invalid.
  message: string
} & (
  | {
      // parsed as JSON
      readonly _tag: 'parsed'

      // e.g., `P1012`
      errorCode?: string
    }
  | {
      // text
      readonly _tag: 'unparsed'
    }
)

export class GetConfigError extends Error {
  constructor(params: GetConfigErrorInit) {
    const constructedErrorMessage = match(params)
      .with({ _tag: 'parsed' }, ({ errorCode, message, reason }) => {
        const errorCodeMessage = errorCode ? `Error code: ${errorCode}` : ''
        return `${reason}
${errorCodeMessage}
${message}`
      })
      .with({ _tag: 'unparsed' }, ({ message, reason }) => {
        const detailsHeader = chalk.red.bold('Details:')
        return `${reason}
${detailsHeader} ${message}`
      })
      .exhaustive()

    const errorMessageWithContext = `${constructedErrorMessage}
[Context: getConfig]`

    super(addVersionDetailsToErrorMessage(errorMessageWithContext))
  }
}

export function getEffectiveUrl(ds: DataSource): EnvValue {
  if (ds.directUrl !== undefined) return ds.directUrl

  return ds.url
}

export async function getConfig(options: GetConfigOptions): Promise<ConfigMetaFormat> {
  const cliEngineBinaryType = getCliQueryEngineBinaryType()

  // always use Wasm for now. At some point we'll get rid of the node-api and binary getConfig
  const data: ConfigMetaFormat = await match(cliEngineBinaryType)
    .when(
      () => true,
      () => {
        return getConfigWasm(options)
      },
    )

    /**
     * @deprecated
     */
    .with(BinaryType.libqueryEngine, () => {
      return getConfigNodeAPI(options)
    })
    .with(BinaryType.queryEngine, () => {
      return getConfigBinary(options)
    })
    .exhaustive()
  return data
}

async function getConfigWasm(options: GetConfigOptions) {
  const debugErrorType = createDebugErrorType(debug, 'getConfigWasm')
  debug(`Using getConfig Wasm`)

  const configEither = pipe(
    E.tryCatch(
      () => {
        if (process.env.FORCE_PANIC_QUERY_ENGINE_GET_CONFIG) {
          debug('Triggering a Rust panic...')
          prismaFmt.debug_panic()
        }

        const params = JSON.stringify({
          prismaSchema: options.datamodel,
          datasourceOverrides: {},
          ignoreEnvVarErrors: options.ignoreEnvVarErrors ?? false,
          env: process.env,
        })

        const data = prismaFmt.get_config(params)
        return data
      },
      (e) => ({
        type: 'wasm-error' as const,
        reason: '(get-config wasm)',
        error: e as Error,
      }),
    ),
    E.map((result) => ({ result })),
    E.chainW(({ result }) =>
      // NOTE: this should never fail, as we expect returned values to be valid JSON-serializable strings
      E.tryCatch(
        () => JSON.parse(result) as ConfigMetaFormat,
        (e) => ({
          type: 'parse-json' as const,
          reason: 'Unable to parse JSON',
          error: e as Error,
        }),
      ),
    ),
  )

  if (E.isRight(configEither)) {
    debug('config data retrieved without errors in getConfig Wasm')
    const { right: data } = configEither
    return Promise.resolve(data)
  }

  /**
   * Check which error to throw.
   */
  const error = match(configEither.left)
    .with({ type: 'wasm-error' }, (e) => {
      debugErrorType(e)

      /**
       * Capture and propagate possible Wasm panics.
       */
      if (isWasmPanic(e.error)) {
        const wasmError = e.error as Error
        const panic = new RustPanic(
          /* message */ wasmError.message,
          /* rustStack */ wasmError.stack || 'NO_BACKTRACE',
          /* request */ '@prisma/prisma-fmt-wasm get_config',
          ErrorArea.FMT_CLI,
          /* schemaPath */ options.prismaPath,
          /* schema */ options.datamodel,
        )
        return panic
      }

      const errorOutput = e.error.message
      return parseConfigError({ errorOutput, reason: e.reason })
    })
    .otherwise((e) => {
      debugErrorType(e)
      return new GetConfigError({ _tag: 'unparsed', message: e.error.message, reason: e.reason })
    })

  throw error
}

async function getConfigNodeAPI(options: GetConfigOptions) {
  const debugErrorType = createDebugErrorType(debug, 'getConfigNodeAPI')

  /**
   * Perform side effects to retrieve variables and metadata that may be useful in the main pipeline's
   * error handling.
   */
  const preliminaryEither = await preliminaryNodeAPIPipeline(options)()
  if (E.isLeft(preliminaryEither)) {
    const { left: e } = preliminaryEither
    debugErrorType(e)
    throw new GetConfigError({ _tag: 'unparsed', message: e.error.message, reason: e.reason })
  }
  const { queryEnginePath } = preliminaryEither.right
  debug(`Using CLI Query Engine (Node-API Library) at: ${queryEnginePath}`)

  /**
   * - load the query engine library
   * - run the "getConfig" command
   */
  const pipeline = pipe(
    loadNodeAPILibrary(queryEnginePath),
    TE.chainW(({ NodeAPIQueryEngineLibrary }) => {
      debug('Loaded Node-API Library')
      return TE.tryCatch(
        () => {
          if (process.env.FORCE_PANIC_QUERY_ENGINE_GET_CONFIG) {
            debug('Triggering a Rust panic...')
            return NodeAPIQueryEngineLibrary.debugPanic('FORCE_PANIC_QUERY_ENGINE_GET_CONFIG')
          }

          // TODO: `result` is a `ConfigMetaFormat`, even though it's typed as `Promise<ConfigMetaFormat>`, so we should
          // probably change the type definition in `Library.ts` (in `@prisma/engine-core`)
          const data = NodeAPIQueryEngineLibrary.getConfig({
            datamodel: options.datamodel,
            datasourceOverrides: {},
            ignoreEnvVarErrors: options.ignoreEnvVarErrors ?? false,
            env: process.env,
          })
          return Promise.resolve(data)
        },
        (e) => ({
          type: 'node-api' as const,
          reason: '(query-engine-node-api library)',
          error: e as Error,
        }),
      )
    }),
  )

  const configEither = await pipeline()
  if (E.isRight(configEither)) {
    debug('config data retrieved without errors in getConfigNodeAPI')
    const { right: data } = configEither
    return data
  }

  /**
   * Check which error to throw.
   */
  const error = match(configEither.left)
    .with({ type: 'node-api' }, (e) => {
      debugErrorType(e)

      /*
       * Extract the actual error by attempting to JSON-parse the error message.
       */
      const errorOutput = e.error.message
      const actualError = pipe(
        E.tryCatch(
          () => JSON.parse(errorOutput),
          () => {
            debug(`Couldn't apply JSON.parse to "${errorOutput}"`)
            return new GetConfigError({ _tag: 'unparsed', message: errorOutput, reason: e.reason })
          },
        ),
        E.map((errorOutputAsJSON: Record<string, string>) => {
          if (errorOutputAsJSON.is_panic) {
            const panic = new RustPanic(
              /* message */ errorOutputAsJSON.message,
              /* rustStack */ errorOutputAsJSON.backtrace || e.error.stack || 'NO_BACKTRACE',
              /* request */ 'query-engine-node-api get-config', // TODO: understand which type it expects
              ErrorArea.QUERY_ENGINE_LIBRARY_CLI,
              /* schemaPath */ options.prismaPath,
              /* schema */ undefined,
            )
            debug(`panic in getConfigNodeAPI "${e.type}"`, panic)
            return panic
          }

          const { error_code: errorCode } = errorOutputAsJSON as { error_code: string | undefined }

          return new GetConfigError({
            _tag: 'parsed',
            message: errorOutputAsJSON.message,
            reason: createSchemaValidationError(e.reason),
            errorCode,
          })
        }),
        E.getOrElseW(identity),
      )

      return actualError
    })
    .otherwise((e) => {
      debugErrorType(e)
      return new GetConfigError({ _tag: 'unparsed', message: e.error.message, reason: e.reason })
    })

  throw error
}

async function getConfigBinary(options: GetConfigOptions) {
  const debugErrorType = createDebugErrorType(debug, 'getConfigBinary')

  /**
   * Perform side effects to retrieve variables and metadata that may be useful in the main pipeline's
   * error handling. For instance, `tempDatamodelPath` is used when submit a Rust panic error report.
   */
  const preliminaryEither = await preliminaryBinaryPipeline(options)()
  if (E.isLeft(preliminaryEither)) {
    const { left: e } = preliminaryEither
    debugErrorType(e)
    throw new GetConfigError({ _tag: 'unparsed', message: e.error.message, reason: e.reason })
  }
  const { queryEnginePath, tempDatamodelPath } = preliminaryEither.right
  debug(`Using CLI Query Engine (Binary) at: ${queryEnginePath}`)
  debug(`PRISMA_DML_PATH: ${tempDatamodelPath}`)

  /**
   * - run the `query-engine cli get-config` command
   * - JSON-deserialize the command output
   */
  const pipeline = pipe(
    (() => {
      const execaOptions = {
        cwd: options.cwd,
        env: {
          /**
           * NOTE: we could potentially Base64-encode the datamodel and store it the `PRISMA_DML` environment variable,
           * which is alternative to `PRISMA_DML_PATH`, but then chances are we'd get an `E2BIG` error when trying
           * to serialize big datamodels.
           */
          PRISMA_DML_PATH: tempDatamodelPath,
          RUST_BACKTRACE: '1',
        },
        maxBuffer: MAX_BUFFER,
      }

      const engineArgs = []
      const args = options.ignoreEnvVarErrors ? ['--ignoreEnvVarErrors'] : []

      return TE.tryCatch(
        () => {
          if (process.env.FORCE_PANIC_QUERY_ENGINE_GET_CONFIG) {
            debug('Triggering a Rust panic...')
            return execa(
              queryEnginePath,
              [...engineArgs, 'cli', 'debug-panic', '--message', 'FORCE_PANIC_QUERY_ENGINE_GET_CONFIG'],
              execaOptions,
            )
          }

          return execa(queryEnginePath, [...engineArgs, 'cli', 'get-config', ...args], execaOptions)
        },
        (e) => ({
          type: 'execa' as const,
          reason: '(query-engine binary)',
          error: e as execa.ExecaError,
        }),
      )
    })(),
    TE.map((result) => ({ result })),
    TE.chainW(({ result }) => {
      return pipe(
        E.tryCatch(
          () => JSON.parse(result.stdout) as ConfigMetaFormat,
          (e) => ({
            type: 'parse-json' as const,
            reason: 'Unable to parse JSON',
            error: e as Error,
          }),
        ),
        TE.fromEither,
      )
    }),
  )

  const configEither = await pipeline()

  if (E.isRight(configEither)) {
    debug('config data retrieved without errors in getConfigBinary')

    // Unlink only when no error occurs, as `sendPanic` might need the `tempDatamodelPath` later in case of errors.
    await unlinkTempDatamodelPath(options, tempDatamodelPath)()

    const { right: data } = configEither
    return data
  }

  /**
   * Check which error to throw.
   */
  const error: RustPanic | GetConfigError = match(configEither.left)
    .with({ type: 'execa' }, (e) => {
      debugErrorType(e)

      /**
       * Capture and propagate possible Rust panics.
       */
      if (isExecaErrorCausedByRustPanic(e.error)) {
        const panic = new RustPanic(
          /* message */ e.error.shortMessage,
          /* rustStack */ e.error.stderr,
          /* request */ 'query-engine get-config',
          ErrorArea.QUERY_ENGINE_BINARY_CLI,
          /* schemaPath */ options.datamodelPath ?? tempDatamodelPath,
          /* schema */ undefined,
        )
        debug(`panic in getConfigBinary "${e.type}"`, panic)
        return panic
      }

      /**
       * Extract the actual error by attempting to JSON-parse the output of the query-engine binary.
       */
      const errorOutput = e.error.stderr ?? e.error.stdout
      return parseConfigError({ errorOutput, reason: e.reason })
    })
    .otherwise((e) => {
      debugErrorType(e)
      return new GetConfigError({ _tag: 'unparsed', message: e.error.message, reason: e.reason })
    })

  throw error
}

type ParseConfigError = {
  errorOutput: string
  reason: string
}

function parseConfigError({ errorOutput, reason }: ParseConfigError): GetConfigError {
  const actualError = pipe(
    E.tryCatch(
      () => JSON.parse(errorOutput),
      () => {
        debug(`Coudln't apply JSON.parse to "${errorOutput}"`)
        return new GetConfigError({ _tag: 'unparsed', message: errorOutput, reason })
      },
    ),
    E.map((errorOutputAsJSON: Record<string, string>) => {
      const defaultMessage = chalk.redBright(errorOutputAsJSON.message)
      const getConfigErrorInit = match(errorOutputAsJSON)
        .with({ error_code: 'P1012' }, (eJSON) => {
          return {
            reason: createSchemaValidationError(reason),
            errorCode: eJSON.error_code,
          }
        })
        .with({ error_code: P.string }, (eJSON) => {
          return {
            reason,
            errorCode: eJSON.error_code,
          }
        })
        .otherwise(() => {
          return {
            reason,
          }
        })

      return new GetConfigError({ _tag: 'parsed', message: defaultMessage, ...getConfigErrorInit })
    }),
    E.getOrElse(identity),
  )
  return actualError
}
