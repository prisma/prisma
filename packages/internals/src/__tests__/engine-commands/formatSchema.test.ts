import execa from 'execa'
import fs from 'fs'
import path from 'path'
import stripAnsi from 'strip-ansi'

import { BinaryType, jestConsoleContext, jestContext, resolveBinary } from '../..'
import { formatSchema } from '../../engine-commands'
import { fixturesPath } from '../__utils__/fixtures'

if (process.env.CI) {
  jest.setTimeout(20_000)
}

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

async function formatSchemaBinary(schema: string): Promise<string> {
  const MAX_BUFFER = 1_000_000_000
  const prismaFmtPath = await resolveBinary(BinaryType.prismaFmt)

  const options = {
    env: {
      RUST_BACKTRACE: process.env.RUST_BACKTRACE ?? '1',
    },
    maxBuffer: MAX_BUFFER,
  } as execa.Options

  const formattedSchema = await execa(prismaFmtPath, ['format'], {
    ...options,
    input: schema,
  })

  return formattedSchema.stdout
}

describe('format wasm vs binary', () => {
  describe('diff', () => {
    test('2-spaces', async () => {
      const schema = await fs.promises.readFile(path.join(fixturesPath, 'format', 'schema-2-spaces.prisma'), {
        encoding: 'utf8',
      })
      const formattedByBinary = await formatSchemaBinary(schema)
      const formattedByWasm = await formatSchema({ schema })
      expect(formattedByBinary + '\n').toEqual(formattedByWasm)
    })

    test('2-spaces-as-tab', async () => {
      const schema = await fs.promises.readFile(path.join(fixturesPath, 'format', 'schema-2-spaces-as-tab.prisma'), {
        encoding: 'utf8',
      })
      const formattedByBinary = await formatSchemaBinary(schema)
      const formattedByWasm = await formatSchema({ schema })
      expect(formattedByBinary + '\n').toEqual(formattedByWasm)
    })

    test('4-spaces', async () => {
      const schema = await fs.promises.readFile(path.join(fixturesPath, 'format', 'schema-4-spaces.prisma'), {
        encoding: 'utf8',
      })
      const formattedByBinary = await formatSchemaBinary(schema)
      const formattedByWasm = await formatSchema({ schema })
      expect(formattedByBinary + '\n').toEqual(formattedByWasm)
    })

    test('4-spaces-as-tab', async () => {
      const schema = await fs.promises.readFile(path.join(fixturesPath, 'format', 'schema-4-spaces-as-tab.prisma'), {
        encoding: 'utf8',
      })
      const formattedByBinary = await formatSchemaBinary(schema)
      const formattedByWasm = await formatSchema({ schema })
      expect(formattedByBinary + '\n').toEqual(formattedByWasm)
    })
  })
})

describe('format custom options', () => {
  const schema = `
  datasource db {
   provider = "sqlite"
   url  = "file:dev.db"
  }
  
  model User {
   id String @default(cuid()) @id
   email String @unique
   name String?
   posts Post[]
  }
`

  test('tabSize=2', async () => {
    const formatted = await formatSchema({ schema }, { tabSize: 2 })
    expect(formatted).toMatchInlineSnapshot(`
      "datasource db {
        provider = "sqlite"
        url      = "file:dev.db"
      }

      model User {
        id    String  @id @default(cuid())
        email String  @unique
        name  String?
        posts Post[]
      }
      "
    `)
  })

  test('tabSize=4', async () => {
    const formatted = await formatSchema({ schema }, { tabSize: 4 })
    expect(formatted).toMatchInlineSnapshot(`
      "datasource db {
          provider = "sqlite"
          url      = "file:dev.db"
      }

      model User {
          id    String  @id @default(cuid())
          email String  @unique
          name  String?
          posts Post[]
      }
      "
    `)
  })
})

describe('format', () => {
  test('nothing', async () => {
    try {
      // @ts-expect-error
      await formatSchema({})
    } catch (e) {
      expect(e.message).toMatchSnapshot()
    }
  })

  test('valid blog schemaPath', async () => {
    const formatted = await formatSchema({
      schemaPath: path.join(fixturesPath, 'blog.prisma'),
    })

    expect(formatted).toMatchSnapshot()
  })

  test('valid blog schema', async () => {
    const formatted = await formatSchema({
      schema: `
      datasource db {
        provider = "sqlite"
        url      = "file:dev.db"
      }
      
      generator client {
        provider      = "prisma-client-js"
        binaryTargets = ["native"]
      }
      
      model User {
        id    String  @default(cuid()) @id
        email String  @unique
        name  String?
        posts Post[]
      }
      
      model Post {
        id        String   @default(cuid()) @id
        createdAt DateTime @default(now())
        updatedAt DateTime @updatedAt
        published Boolean
        title     String
        content   String?
        authorId  String?
        author    User?    @relation(fields: [authorId], references: [id])
      }
      
      model Like {
        id     String @default(cuid()) @id
        userId String
        user   User   @relation(fields: [userId], references: [id])
        postId String
        post   Post   @relation(fields: [postId], references: [id])
      
        @@unique([userId, postId])
      }`,
    })

    expect(formatted).toMatchSnapshot()
  })

  test('valid schema with 1 preview feature flag warning', async () => {
    const schema = /* prisma */ `
      generator client {
        provider = "prisma-client-js"
        previewFeatures = ["cockroachdb"]
      }

      datasource db {
        provider = "cockroachdb"
        url = env("TEST_POSTGRES_URI")
      }

      model SomeUser {
        id   Int    @id
      }
    `
    const formattedSchema = await formatSchema({ schema })
    expect(formattedSchema).toMatchSnapshot()

    expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
    expect(stripAnsi(ctx.mocked['console.warn'].mock.calls.join('\n'))).toMatchInlineSnapshot(`
      "
      Prisma schema warning:
      - Preview feature "cockroachdb" is deprecated. The functionality can be used without specifying it as a preview feature."
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('valid schema with 3 preview feature flag warnings', async () => {
    const schema = /* prisma */ `
      generator client {
        provider = "prisma-client-js"
        previewFeatures = ["cockroachdb", "mongoDb", "microsoftSqlServer"]
      }

      datasource db {
        provider = "cockroachdb"
        url = env("TEST_POSTGRES_URI")
      }

      model SomeUser {
        id   Int    @id
      }
    `
    const formattedSchema = await formatSchema({ schema })
    expect(formattedSchema).toMatchSnapshot()

    expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
    expect(stripAnsi(ctx.mocked['console.warn'].mock.calls.join('\n'))).toMatchInlineSnapshot(`
      "
      Prisma schema warnings:
      - Preview feature "cockroachdb" is deprecated. The functionality can be used without specifying it as a preview feature.
      - Preview feature "mongoDb" is deprecated. The functionality can be used without specifying it as a preview feature.
      - Preview feature "microsoftSqlServer" is deprecated. The functionality can be used without specifying it as a preview feature."
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('invalid schema', async () => {
    const schema = /* prisma */ `
      generator client {
        provider        = "prisma-client-js"
      }

      datasource db {
        provider = "cockroachdb"
        url      = env("TEST_POSTGRES_URI")
      }

      model SomeUser {
        id      Int      @id
        profile Profile?
      }

      model Profile {
        id     Int      @id
        user   SomeUser @relation(fields: [userId], references: [id], onUpdate: SetNull)
        userId Int      @unique
      }
    `
    const formattedSchema = await formatSchema({ schema })
    expect(formattedSchema).toMatchInlineSnapshot(`
      "generator client {
        provider = "prisma-client-js"
      }

      datasource db {
        provider = "cockroachdb"
        url      = env("TEST_POSTGRES_URI")
      }

      model SomeUser {
        id      Int      @id
        profile Profile?
      }

      model Profile {
        id     Int      @id
        user   SomeUser @relation(fields: [userId], references: [id], onUpdate: SetNull)
        userId Int      @unique
      }
      "
    `)

    expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
    expect(ctx.mocked['console.warn'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })
})
