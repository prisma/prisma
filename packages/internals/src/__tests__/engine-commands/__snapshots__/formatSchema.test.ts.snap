// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format nothing 1`] = `"Parameter schema or schemaPath must be passed."`;

exports[`format valid blog schema 1`] = `
"datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native"]
}

model User {
  id    String  @id @default(cuid())
  email String  @unique
  name  String?
  posts Post[]
  Like  Like[]
}

model Post {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean
  title     String
  content   String?
  authorId  String?
  author    User?    @relation(fields: [authorId], references: [id])
  Like      Like[]
}

model Like {
  id     String @id @default(cuid())
  userId String
  user   User   @relation(fields: [userId], references: [id])
  postId String
  post   Post   @relation(fields: [postId], references: [id])

  @@unique([userId, postId])
}
"
`;

exports[`format valid blog schemaPath 1`] = `
"datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id    String  @id @default(cuid())
  email String  @unique
  name  String?
  posts Post[]
  Like  Like[]
}

model Post {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean
  title     String
  content   String?
  authorId  String?
  author    User?    @relation(fields: [authorId], references: [id])
  Like      Like[]
}

model Like {
  id     String @id @default(cuid())
  userId String
  user   User   @relation(fields: [userId], references: [id])
  postId String
  post   Post   @relation(fields: [postId], references: [id])

  @@unique([userId, postId])
}
"
`;

exports[`format valid schema with 1 preview feature flag warning 1`] = `
"generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["cockroachdb"]
}

datasource db {
  provider = "cockroachdb"
  url      = env("TEST_POSTGRES_URI")
}

model SomeUser {
  id Int @id
}
"
`;

exports[`format valid schema with 3 preview feature flag warnings 1`] = `
"generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["cockroachdb", "mongoDb", "microsoftSqlServer"]
}

datasource db {
  provider = "cockroachdb"
  url      = env("TEST_POSTGRES_URI")
}

model SomeUser {
  id Int @id
}
"
`;
